pkg/registry/core/service/strategy.go

```diff
@@ -205,6 +207,127 @@ func topologyKeysInUse(svc *api.Service) bool {
+// fields that are allocated from a pool and need to be released.  Anyone who
+// is contemplating copying this pattern should think REALLY hard about almost
+// any other option.
+func dropTypeDependentFields(newSvc *api.Service, oldSvc *api.Service) {
+	// For now we are only normalizing on updates.  This minimizes potential
+	// confusion since many of the cases we are handling here are pretty niche.
+	if oldSvc == nil {
+		return
+	}
+
+	// If the user is switching to a type that does not need a value (even
+	// "None" counts), AND they did not change the ClusterIP value, we can wipe
+	// it.  It will be deallocated later.
+	//
+	// There are other fields that could also be wiped, but historically were
+	// ignored.  For compat, we leave it that way.
+	if needsClusterIP(oldSvc) && !needsClusterIP(newSvc) && sameClusterIP(oldSvc, newSvc) {
+		newSvc.Spec.ClusterIP = ""
+	}
+
+	// If the user is switching to a type that doesn't use NodePorts AND they
+	// did not change any NodePort values, we can wipe them.  It will be
+	// deallocated later.
+	if needsNodePort(oldSvc) && !needsNodePort(newSvc) && sameNodePorts(oldSvc, newSvc) {
+		for i := range newSvc.Spec.Ports {
+			newSvc.Spec.Ports[i].NodePort = 0
+		}
+	}
+
+	// If the user is switching to a case that doesn't use HealthCheckNodePort AND they
+	// did not change the HealthCheckNodePort value, we can wipe it.  It will
+	// be deallocated later.
+	if needsHCNodePort(oldSvc) && !needsHCNodePort(newSvc) && sameHCNodePort(oldSvc, newSvc) {
+		newSvc.Spec.HealthCheckNodePort = 0
+	}
+
+	// NOTE: there are other fields like `selector` which we could wipe.
+	// Historically we did not wipe them and they are not allocated from
+	// finite pools, so we are (currently) choosing to leave them alone.
+}
+
+func needsClusterIP(svc *api.Service) bool {
+	if svc.Spec.Type == api.ServiceTypeExternalName {
+		return false
+	}
+	return true
+}
+
+func sameClusterIP(oldSvc, newSvc *api.Service) bool {
+	return oldSvc.Spec.ClusterIP == newSvc.Spec.ClusterIP
+}
+
+func needsNodePort(svc *api.Service) bool {
+	if svc.Spec.Type == api.ServiceTypeNodePort || svc.Spec.Type == api.ServiceTypeLoadBalancer {
+		return true
+	}
+	return false
+}
+
+func sameNodePorts(oldSvc, newSvc *api.Service) bool {
+	// Helper to make a unique key so we can tell if a port changed.
+	makeKey := func(p *api.ServicePort) string {
+		return fmt.Sprintf("%s/%d", p.Protocol, p.Port)
+	}
+	// Helper to make a map of port key -> NodePort value.
+	mapNodePorts := func(svc *api.Service) map[string]int32 {
+		out := make(map[string]int32)
+		for i := range svc.Spec.Ports {
+			key := makeKey(&svc.Spec.Ports[i])
+			out[key] = svc.Spec.Ports[i].NodePort
+		}
+		return out
+	}
+
+	oldPorts := mapNodePorts(oldSvc)
+	newPorts := mapNodePorts(newSvc)
+
+	// Every old port must have the same NodePort in the new config or be
+	// cleared to 0 or must be removed.
+	for k, oldVal := range oldPorts {
+		if newVal, found := newPorts[k]; found {
```

Comment:
What happens to partial remove, eg change type and remove one of n ports?

Reply:
If they remove a port it is fine.

If they change a port number or protocol, this treats it as if they removed the old and added the new.  If they did not clear NodePort, we will not clear it (they "added" a port with NodePort set) and will fail validation.

I thought it was important to allow adding a new port and removing an old port.  Changing an existing port is more complicated.  I could buy an argument that as long as the node port value didn't change, we could still wipe it (though that's not what is here).  Do you think that's better?  Now that I spell it out, I am coming around..

examples (assume {protocol, port, nodeport)


Good Add: New port does not have a NodePort
`{ type: NodePort, ports: [{TCP, 80, 30080}] }` -> `{ type: ClusterIP, ports: [{TCP, 80, 30080}, {TCP, 443}] }`

Bad Add: New port has a NodePort
`{ type: NodePort, ports: [{TCP, 80, 30080}] }` -> `{ type: ClusterIP, ports: [{TCP, 80, 30080}, {TCP, 443, 30443}] }`

Good Del:
`{ type: NodePort, ports: [{TCP, 80, 30080}, {TCP, 443, 30443}] }` -> `{ type: ClusterIP, ports: [{TCP, 80, 30080}] }`

Good Mod: New port does not have a NodePort
`{ type: NodePort, ports: [{TCP, 80, 30080} }` -> `{ type: ClusterIP, ports: [{TCP, 443}] }`

The open question is this one (set to demonstrate worst case):
`{ type: NodePort, ports: [{TCP, 80, 30080} }` -> `{ type: ClusterIP, ports: [{UDP, 53, 30080}] }`

Both the protocol and port number changed, but the NodePort stayed the same.  Can I wipe the node port safely?  I was thinking not, but now I am changing my thinking...

Reply:
New push is up as a 2nd commit.  Will mark this PR for squash

Reply:
the original line of thought (or at least that is what i thought we are trying to do):
as captured by this line https://github.com/kubernetes/kubernetes/pull/95196/files#diff-2dd9a1ef6872092e519941c8a4870f4c2a71eda096b660510b02afe9b3d19fefR243

`user should be able to change a service type to a service that does not need NodePorts without having to wipe out NodePorts while patching the Service`

**a truth table would look like this** 
O = old
N = new
O-type-need-node-ports => N-type-need-node-ports + ports changed               `OK: if node port is immutable*`
O-type-need-node-ports => N-type-need-node-ports + ports same                     `OK: validate other fields`
O-type-need-node-ports => N-DO NOT-need-node-ports + ports same               `OK: clear intent`
O-type-need-node-ports => N-DO NOT-need-node-ports + ports cleared            `OK: clear intent`
O-type-need-node-ports => N-DO NOT-need-node-ports + ports changed `NOT OK: unclear intent **`
O-type-DO NOT- need-node-ports => N-DO NOT-need-node-ports + add ports `NOT OK: ports! (validation)`
O-type-DO NOT- need-node-ports => N-DO NOT-need-node-ports + no ports    `OK`
O-type-DO NOT- need-node-ports => N-need-node-ports + no ports                    `OK`
O-type-DO NOT- need-node-ports => N-need-node-ports + some ports `OK: if ports available (alloc)***`
O-type-DO NOT- need-node-ports => N-need-node-ports + all ports `OK: if ports available (alloc)***`

* i am not even sure if we allow changing them (i am assuming yes)
** any change add/change/partial delete did the user really want to change type or made a mistake?
*** assuming we allow specific port allocation if available (similar to clusterIPs)

Reply:
I think it's safe to assume that a user who changed ports meant to do so.  E.g.

CREATE { type=LB, ports[0] = (TCP, 8080) }
PATCH { type=ClusterIP, ports[0]=(TCP, 8000) }

I think that should NOT cause an error - we need to clear the NodePort.